translators/src/Print/Idris.hs

```diff
@@ -1,110 +1,155 @@
+printType (PCon name types) = pretty name <+> hsep (map printType types)
+printType (DCon name [] exprs) = pretty name <+> hsep (map printExpr exprs)
 printType (DCon name types exprs) = -- For dependent type constructors
-    name ++ " " ++ unwords (map printType types) ++ " " ++ unwords (map printExpr exprs)
-printType (Index names ty) = brackets $ intercalate ", " names ++ " : " ++ printType ty
+    pretty name <+> hsep (map printType types) <+> hsep (map printExpr exprs)
+printType (Index names ty) = braces $ typeAnn (hsep $ punctuate comma $ map pretty names) (printType ty)
 printType (Embed e) = printExpr e
 
-printExpr :: Expr -> String
-printExpr (Constructor name) = name
-printExpr (Var var) = var
-printExpr (Nat n) = show n
-printExpr (String str) = quote str
+
+printExpr :: Expr -> Doc ann
+printExpr (Constructor name) = pretty name
+printExpr (Var var) = pretty var
+printExpr (Nat n) = pretty n
+printExpr (String str) = dquotes $ pretty str
 printExpr (Paren e) = parens $ printExpr e
-printExpr (Bin op e1 e2) = printExpr e1 ++ printOp op ++ printExpr e2
-printExpr (Let [] expr) = printExpr expr -- this should never happen
-printExpr (Let (d:[]) expr) = "let \n    " ++ (printDef d) ++ " in \n    " ++ printExpr expr
-printExpr (Let (d:ds) expr) = 
-  "let \n    " ++ intercalate "\n    " (map printDef (d:ds)) ++
-  "\n    in \n    " ++ printExpr expr -- probably need to recursively indent blocks to make sure everything stays aligned
-printExpr (If cond thn els) = "if " ++ printExpr cond ++ " then " ++ printExpr thn ++ " else " ++ printExpr els
-printExpr (Where expr ds) = printExpr expr ++ "\n    where " ++ intercalate "\n    " (map printDef ds)
-printExpr (FunCall fun args) = fun ++ " " ++ (unwords $ map printExpr args) -- Added case for FunCall
-printExpr (VecE l) = sqbrackets $ intercalate ", " (map printExpr l)
-printExpr (ListE l) = sqbrackets $ intercalate ", " (map printExpr l)
-printExpr (Suc t) = parens $ "S " ++ printExpr t
-
-printOp :: Op -> String
-printOp Plus = " + "
-
-
-printDef :: Definition -> String
-printDef (DefUVar var expr) = var ++ assign ++ printExpr expr
-printDef (DefTVar var t expr) = line (var ++ typedel ++ printType t) ++ var ++ assign ++ printExpr expr
-printDef (DefFun var ty args expr) = typeSig ++ var ++ " " ++ (intercalate " " $ map arg args) ++ assign ++ printExpr expr
-    where typeSig = maybe "" (\t -> var ++ typedel ++ printType t ++ "\n    ") ty
-printDef (DefNesFun var Nothing args expr) = printDef (DefFun var Nothing args expr)
-printDef (DefNesFun var (Just t) args expr) = printDef (DefFun var (Just t) args expr)
+printExpr (Bin op e1 e2) = printExpr e1 <+> printOp op <+> printExpr e2
+printExpr (Let ds expr) = 
+  "let" <+> align (vcat (map printLocalDefn ds) <+> "in") <> line <>
+  printExpr expr
+printExpr (If cond thn els) =
+  "if" <+> printExpr cond <+> "then" <+> printExpr thn <+> "else" <+> printExpr els
+printExpr (Where expr ds) =
+  printExpr expr <> line <>
```

Comment:
This should be a `hardline` I think?

Also, the pattern of `hardline <> indent n doc` seems to be pretty common; might be worth factoring out.

Reply:
There are quite a few patterns emerging. I'm waiting for Lean to be done, and then going to do a pass of factoring.